In [1]:
# Importing required librries

import pandas as pd
from sklearn.metrics import accuracy_score
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import GridSearchCV

In [2]:
# Importing the data

dataset = pd.read_csv('news.csv')
dataset.head()

Unnamed: 0                                              title  \
0        8476                       You Can Smell Hillary’s Fear   
1       10294  Watch The Exact Moment Paul Ryan Committed Pol...   
2        3608        Kerry to go to Paris in gesture of sympathy   
3       10142  Bernie supporters on Twitter erupt in anger ag...   
4         875   The Battle of New York: Why This Primary Matters   

                                                text label  
0  Daniel Greenfield, a Shillman Journalism Fello...  FAKE  
1  Google Pinterest Digg Linkedin Reddit Stumbleu...  FAKE  
2  U.S. Secretary of State John F. Kerry said Mon...  REAL  
3  — Kaydee King (@KaydeeKing) November 9, 2016 T...  FAKE  
4  It's primary day in New York and front-runners...  REAL

In [3]:
y = dataset.iloc[:,3:4]

In [4]:
# Importing libraries for cleaning the data

import re
import nltk
from nltk.corpus import stopwords
from nltk.stem.porter import PorterStemmer
corpus = []

In [5]:
# Cleaning the data

for i in range(0,6335):
    review = re.sub('[^a-zA-Z]',' ',dataset['text'][i])
    review = review.lower()
    review = review.split()
    ps = PorterStemmer()
    review = [ps.stem(word) for word in review if not word in set(stopwords.words('english'))]
    review = ' '.join(review)
    corpus.append(review)

In [6]:
# Vectorizing our data

from sklearn.feature_extraction.text import CountVectorizer
cv = CountVectorizer()
X = pd.DataFrame(cv.fit_transform(corpus).toarray())

In [7]:
# Spltiing our data into training and test set

from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test = train_test_split(X,y,train_size = 0.8,random_state=56)

In [9]:
# Using logistic regression as our classification model

from sklearn.linear_model import LogisticRegression
classifier = LogisticRegression(random_state=68,solver = 'liblinear')
classifier.fit(X_train,y_train)
y_pred_logistic_regression = classifier.predict(X_test)
print('Accuracy for test set: '+ str(accuracy_score(y_test,y_pred_logistic_regression)))
y_pred_logistic_regression_train = classifier.predict(X_train)
print('Accuracy for train set: '+ str(accuracy_score(y_train,y_pred_logistic_regression_train)))
acc = cross_val_score(estimator = classifier,X = X_train,y = y_train,cv = 10 ,scoring = 'accuracy')
print('CV_Accuracy: ' + str(acc.mean()) + ' +/- ' + str(acc.std()))

C:\Users\utkar\Anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)


Accuracy for test set: 0.9155485398579322
Accuracy for train set: 0.999802683504341


C:\Users\utkar\Anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\utkar\Anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\utkar\Anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), for example using ravel().
  y = column_or_1d(y, warn=True)
C:\Users\utkar\Anaconda3\lib\site-packages\sklearn\utils\validation.py:724: DataConversionWarning: A column-vector y was passed when a 1d array was expected. Please change the shape of y to (n_samples, ), 

CV_Accuracy: 0.9092374089479955 +/- 0.010634332374767167


Our model gives an accuracy of 99.98 % on the training set and an accuracy of 91.55 % on our test set. We also get a mean accuracy of 90.92 % using cross-validation.